pkg/virt-handler/virtwrap/isolation/isolation.go

```diff
@@ -0,0 +1,129 @@
+package isolation
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"syscall"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/logging"
+)
+
+type PodIsolationDetector interface {
+	Detect(vm *v1.VM) (IsolationResult, error)
+}
+
+type IsolationResult interface {
+	Controller() []string
+	Slice() string
+	Pid() int
+	PidNS() string
+}
+
+type socketBasedIsolationDetector struct {
+	socketDir string
+}
+
+func NewSocketBasedIsolationDetector(socketDir string) PodIsolationDetector {
+	return &socketBasedIsolationDetector{socketDir: socketDir}
+}
+
+func (s *socketBasedIsolationDetector) Detect(vm *v1.VM) (IsolationResult, error) {
+	var pid int
+	var slice string
+	var err error
+	var controller []string
+
+	socket := s.socketDir + "/" + vm.ObjectMeta.Namespace + "/" + vm.ObjectMeta.Name + "/sock"
+	if pid, err = s.getPid(socket); err != nil {
+		logging.DefaultLogger().Object(vm).Error().Reason(err).V(3).Msgf("Could not get owner Pid of socket %s", socket)
+		return nil, err
+
+	}
+	if controller, slice, err = s.getSlice(pid); err != nil {
+		logging.DefaultLogger().Object(vm).Error().Reason(err).V(3).Msgf("Could not get cgroup slice for Pid %d", pid)
+		return nil, err
+	}
+
+	return &isolationResult{pid: pid, slice: slice, controller: controller}, nil
+}
+
+type isolationResult struct {
+	pid        int
+	slice      string
+	controller []string
+}
+
+func (r *isolationResult) Slice() string {
+	return r.slice
+}
+
+func (r *isolationResult) PidNS() string {
+	return fmt.Sprintf("/proc/%d/ns/pid", r.pid)
+}
+
+func (r *isolationResult) Pid() int {
+	return r.pid
+}
+
+func (r *isolationResult) Controller() []string {
+	return r.controller
+}
+
+func (s *socketBasedIsolationDetector) getPid(socket string) (int, error) {
+	sock, err := net.Dial("unix", socket)
+	if err != nil {
+		return -1, err
+	}
+	defer sock.Close()
+
+	ufile, err := sock.(*net.UnixConn).File()
+	if err != nil {
+		return -1, err
+	}
```

Comment:
I don't think that keeping the FileDescriptor open helps avoiding that race. We would still have already created and started the VM, before we would detect that the file descriptor is now closed. Also see https://github.com/kubevirt/kubevirt/pull/337#issuecomment-316970884 how this fits into the whole picture.